segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/hdfs/HDFSStorage.java

```diff
@@ -196,65 +333,128 @@ public boolean supportsTruncation() {
     }
 
 
-    @Override
-    public int read(SegmentHandle handle, long offset, byte[] buffer, int bufferOffset, int length) throws StreamSegmentException {
-        return call(new ReadOperation(asReadableHandle(handle), offset, buffer, bufferOffset, length, this.context));
-    }
 
     @Override
-    public SegmentProperties getStreamSegmentInfo(String streamSegmentName) throws StreamSegmentException {
-        return call(new GetInfoOperation(streamSegmentName, this.context));
+    public void write(SegmentHandle handle, long offset, InputStream data, int length) throws StreamSegmentException {
+        long traceId = LoggerHelpers.traceEnter(log, "write", handle, offset, length);
+        handle = asWritableHandle(handle);
+        FileStatus lastFile = null;
+        try {
+            lastFile = findStatusForSegment(handle.getSegmentName(), true);
+            if (isSealed(lastFile.getPath())) {
+                throw new StreamSegmentSealedException(handle.getSegmentName());
+            }
+            if (getEpocFromPath(lastFile.getPath()) > this.context.epoch) {
+                throw new StorageNotPrimaryException(handle.getSegmentName());
+            }
+        } catch (IOException e) {
```

Comment:
It's a little weird to be throwing exceptions directly above and rethrowing them in the same method. Can we make the above lines just throw the correct exceptions in the first place?